File: sol/1930.py
from bisect import bisect

class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        orda = ord('a')
        indices = list(list() for _ in range(26))
        for i, c in enumerate(s):
            indices[ord(c) - orda].append(i)
        res = 0
        for i in range(26):
            if len(indices[i]) < 2:
                continue
            if len(indices[i]) > 2:
                res += 1
            for j in range(26):
                if i == j:
                    continue
                if bisect(indices[j], indices[i][0]) < bisect(indices[j], indices[i][-1]):
                    res += 1
        return res

File: sol/1930/1930.cpp
#include <algorithm>

struct Solution {
    int countPalindromicSubsequence(const string &s) {
        vector<int> indices[26];
        for (int i = 0; i < s.size(); ++i){
            indices[s[i] - 'a'].push_back(i);
        }
        int res = 0;
        for (int i = 0; i < 26; ++i){
            if (indices[i].size() < 2){
                continue;
            }
            if (indices[i].size() > 2){
                ++res;
            }
            for (int j = 0; j < 26; ++j){
                if (i == j){
                    continue;
                }
                if (lower_bound(indices[j].cbegin(), indices[j].cend(), indices[i].front())
                        != lower_bound(indices[j].cbegin(), indices[j].cend(), indices[i].back())){
                    ++res;
                }
            }
        }
        return res;
    }
};
